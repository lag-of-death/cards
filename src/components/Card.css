.card-front {
    border-radius: 12px;
    width: 222px;
    height: 315px;
    box-sizing: border-box;
    border: 1px solid black;
    margin: 0 2px 4px;
    background: lightblue;
}
