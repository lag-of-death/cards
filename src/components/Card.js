import React from 'react';
import ReactCardFlip from 'react-card-flip';
import propTypes from 'prop-types';
import { connect } from 'react-redux';

import { actions } from '../store';

function Card({
  card, chooseCard, chosenCards, deckId,
}) {
  const isFlipped = (
    chosenCards.chosen.find(({ code }) => code === card.code + deckId)
      || chosenCards.guessedPairs.find((guessedCode) => guessedCode === card.code)
  );

  return (
    <ReactCardFlip isFlipped={isFlipped} flipDirection="horizontal">
      <button
        type="button"
        onClick={() => chooseCard(card, deckId)}
      >
        {card.code}
      </button>

      <img alt={card.code} src={card.image} />
    </ReactCardFlip>
  );
}

Card.propTypes = {
  chosenCards: propTypes.shape({
    chosen: propTypes.arrayOf(propTypes.shape({ code: propTypes.string })),
    guessedPairs: propTypes.arrayOf(propTypes.string),
  }).isRequired,
  chooseCard: propTypes.func.isRequired,
  deckId: propTypes.string.isRequired,
  card: propTypes.shape({ image: propTypes.string, code: propTypes.string }).isRequired,
};

export default connect(
  ({ chosenCards }) => ({ chosenCards }),
  (dispatch) => ({
    chooseCard: (card, deckId) => {
      dispatch(async (disp, getState, useCases) => {
        const cards = await useCases.chooseCard(card, getState().chosenCards, deckId);

        disp({
          type: actions.CHOOSE_CARD,
          payload: cards,
        });
      });
    },
  }),
)(Card);
