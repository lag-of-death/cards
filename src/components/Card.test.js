import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Card } from './Card';

describe('Card', () => {
  it('calls its callback', () => {
    const cb = jest.fn(() => ({}));
    const cardData = { code: 'AS', image: 'path/to/image' };
    const deckId = 'xyz';

    const card = render(
      <Card
        card={cardData}
        chooseCard={cb}
        chosenCards={{ chosen: [], guessedPairs: [] }}
        deckId={deckId}
      />,
    );

    const cardHandle = card.getByTestId('card');

    fireEvent.click(cardHandle);

    expect(card).toMatchSnapshot();
    expect(cb).toHaveBeenCalledWith(cardData, deckId);
  });
});
