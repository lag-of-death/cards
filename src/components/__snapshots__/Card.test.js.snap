// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card calls its callback 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="react-card-flip"
        style="z-index: auto;"
      >
        <div
          class="react-card-flipper"
          style="position: relative; width: 100%; height: 100%;"
        >
          <div
            class="react-card-front"
            style="left: 0px; position: relative; top: 0px; transform: rotateY(0deg); width: 100%; height: 100%; z-index: 2; transition: 0.6s;"
          >
            <button
              class="card-front"
              data-testid="card"
              type="button"
            >
              AS
            </button>
          </div>
          <div
            class="react-card-back"
            style="left: 0px; position: absolute; transform: rotateY(-180deg); top: 0px; width: 100%; height: 100%; transition: 0.6s;"
          >
            <img
              alt="AS"
              src="path/to/image"
            />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="react-card-flip"
      style="z-index: auto;"
    >
      <div
        class="react-card-flipper"
        style="position: relative; width: 100%; height: 100%;"
      >
        <div
          class="react-card-front"
          style="left: 0px; position: relative; top: 0px; transform: rotateY(0deg); width: 100%; height: 100%; z-index: 2; transition: 0.6s;"
        >
          <button
            class="card-front"
            data-testid="card"
            type="button"
          >
            AS
          </button>
        </div>
        <div
          class="react-card-back"
          style="left: 0px; position: absolute; transform: rotateY(-180deg); top: 0px; width: 100%; height: 100%; transition: 0.6s;"
        >
          <img
            alt="AS"
            src="path/to/image"
          />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
